### PR TITLE
[move-core-types] generalizing MoveStruct to support field names

### PR DESCRIPTION
**Pull request by sblackshear**
Moving toward the direction of having a single MoveStruct/MoveValue implementation that can be used by both the VM (which values compactness) and clients (who value readability). The goal is that most (all?) clients of resource-viewer can use the MoveStructWithFields representation here instead of AnnotatedMoveValue.

### CODE REVIEW

File: language/move-core/types/src/value.rs [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+use std::{
+    collections::BTreeMap,
+    fmt::{self, Debug},
+};
 
 #[derive(Debug, PartialEq, Eq, Clone)]
-pub struct MoveStruct(Vec<MoveValue>);
+pub enum MoveStruct {
+    /// The representation used by the MoveVM
+    Runtime(Vec<MoveValue>),
```
Comment (runtian-zhou) [inline language/move-core/types/src/value.rs:21]: MoveVM won't use this struct in the runtime so I think getting field name by default makes sense here?

**Comment by vgao1996** [comment]
> Moving toward the direction of having a single MoveStruct/MoveValue implementation that can be used by both the VM (which values compactness) and clients (who value readability).

I'm not totally sure if this will work and we should talk. My understanding is that value compactness alone is not a good characterization of the VM's needs. Of course the VM would welcome space saving, but above that, lookup efficiency, or more precisely, constant time lookups are likely more important, sometimes even at the sacrifice of the former. Following this following this line of thought, I think what the VM will eventually want, is a contextualized type representation. By contextualized representation I mean all the actual data is stored in tables, and the type objects themselves are nothing but lightweight pointers that references data in the tables, which cannot be interpreted out of the context.

At meanwhile, there exists tools whose requirements can be met by a simpler, or self contained representation like the one in this PR. In the future, I think it's fine for them to either keep using the simple representation or switch to the contextualized representation, if we really want to unify the representations. I don't think it is possible however, that a simple self-contained representation will ever be able to replace the contextualized one for more complicated applications.

File: language/move-core/types/src/value.rs (hunk 1 above)
Comment (sblackshear) [reply language/move-core/types/src/value.rs:21]: It is currently used in runtime.rs (or at least, `MoveValue` is). Maybe it doesn't need to be used there, but my goal is not cleaning up existing uses; rather, I want to generalize this to allow both representations.

**Comment by sblackshear** [comment]
> By contextualized representation I mean all the actual data is stored in tables, and the type objects themselves are nothing but lightweight pointers that references locations in the tables, which cannot be interpreted out of the context.
> I don't think it is possible however, that a simple self-contained representation will ever be able to replace the contextualized one for more complicated applications.

I agree with this. I think the VM should keep its representation of values internal and have the freedom to change it at will. The self-contained representations should only be for tooling and clients, not seek to serve the VM's needs.

> This could even become a plug-in replacement for the ResourceViewer.

I am indeed hoping that `MoveStruct::WithFields` can be used in many places where we currently use `AnnotatedMoveStructs`.

> One idea to consider: if we move the code cache/loader out of the Move VM, and allow clients to interact with and query it, then it naturally becomes the contextualized representation I mentioned.

I was loooking at this code and thinking about this a few days ago. I'm unsure about this for a few reasons:
- The code cache/loader deal in both `Type`s (which are contextual) and `MoveTypeLayout`'s (which are context-free). The loader needs to get a `Type` before it can get a `MoveTypeLayout`, but clients will care about the latter and not the former. I think it would be better to have a direct route from `TypeTag` to `MoveTypeLayout` that doesn't go through the VM-internal `Type`.
- The code cache/loader have a lot of complexity/dependencies related to caching that I don't think we want or need to inflict on clients. If you're just fetching the layout for one type so you can generate some client bindings for it (likely a common case), populating a cache isn't really worth it.
- I'm wary of touching the perf/safety-critical VM code, but feel much more comfortable writing a simple/direct scheme for deriving `MoveTypeLayout`'s from a collection of modules.

**Comment by bob-wilson** [comment]
I would still like to see a simple loader that abstracts the details of the file format without doing much of anything else. We could then build additional layers on top of that base. I agree with @vgao1996 and @sblackshear about the VM having its own internal representation and something like the proposed "contextualized representation" sounds good. Other tools can have a self-contained representation. That's all good -- I just want to have these share the same code for the basic Move bytecode reading/writing. I've mentioned this in the past and I know that not everyone agrees. Maybe it's time to put some effort into this and try to get a consensus on our roadmap for this....

**Comment by sblackshear** [comment]
> I would still like to see a simple loader that abstracts the details of the file format without doing much of anything else. 

That sounds like a good idea to me. But wondering: how much abstraction are you looking for here--in particular, do you want to expose indexes into the tables?

Another way to frame this might be: is https://github.com/diem/diem/blob/main/language/move-binary-format/src/normalized.rs too abstract for what you want, or not abstract enough?

**Comment by sblackshear** [comment]
> I would still like to see a simple loader that abstracts the details of the file format without doing much of anything else. 

That is basically what this PR is going for at the value and type layout level. And I think https://github.com/diem/diem/blob/main/language/move-binary-format/src/normalized.rs is abstracting struct/function definitions well, but doesn't try to abstract instructions, constants, etc. Would an extension of that approach to instructions be along the lines of what you're looking for?

**Comment by vgao1996** [comment]
Looks like we're mostly on the same page. 

> The code cache/loader deal in both Types (which are contextual) and MoveTypeLayout's (which are context-free). The loader needs to get a Type before it can get a MoveTypeLayout, but clients will care about the latter and not the former. I think it would be better to have a direct route from TypeTag to MoveTypeLayout that doesn't go through the VM-internal Type.
The code cache/loader have a lot of complexity/dependencies related to caching that I don't think we want or need to inflict on clients. If you're just fetching the layout for one type so you can generate some client bindings for it (likely a common case), populating a cache isn't really worth it.

Agreed that the contextualized type representation could be a bit too heavy-weighted for many simple clients. I think it would be great to have it as an option, but we definitely shouldn't feel pressured to use it just for the sake of it. It should probably be reserved for applications with a complexity similar to that of the VM's. For example, I see this as a good fit for Jack's mysql backend. He wouldn't had to reinvent a lot of wheels if we had this.

> I'm wary of touching the perf/safety-critical VM code, but feel much more comfortable writing a simple/direct scheme for deriving MoveTypeLayout's from a collection of modules.

Fair point. Although I think the risk is limited if the clients are just using the existing public APIs offered. It might actually even provide additional testing for this critical component. What could be really dangerous however, is to add additional functionalities used not by the VM, but some other "less critical" components, as this introduces changes and may conflate the design.

Overall, I think Bob's judgement on this is a good one. At least for now, it seems better for the VM and simpler tools to stay on different paths.

File: language/move-core/types/src/value.rs [hunk 2]
```diff
@@ -144,6 +192,38 @@ impl<'d, 'a> serde::de::Visitor<'d> for VectorElementVisitor<'a> {
...
+                        return Err(A::Error::custom(
+                            "Fields ordered differently in layout vs struct",
+                        ));
+                    }
+                    val.insert(other_f, elem);
+                }
+                None => return Err(A::Error::invalid_length(i, &self)),
+            }
+        }
+        Ok(val)
```
Comment (vgao1996) [inline language/move-core/types/src/value.rs:223]: Honestly, this doesn't look quite right to me. Correct me if I'm wrong, but I don't think we are requiring fields to be alphabetically sorted and using a BtreeMap indexed by field names will break the original ordering. A `Vec<(Identifier, MoveValue)>` seems to be the better choice here...

**Review by vgao1996 — DISMISSED** [review]
Honestly, I'm still concerned that this generalization will fracture the clients, creating a situation where one client cannot digest a value generated by another client. 

However on the other hand, I'm starting to like the flexibility this design offers, and I can see it enabling us to do something wasn't possible in the past, like partial annotation, unified value formatter etc. I also just realized that this is similar to JSON in a way, just less flexible. Overall, I think it's worth giving this a pass and try it out in the fields.

File: language/move-core/types/src/value.rs [hunk 3]
```diff
@@ -72,28 +86,62 @@ where
...
+
     pub fn fields(&self) -> &[MoveTypeLayout] {
-        &self.0
+        match self {
+            Self::Runtime(vals) => vals,
+            Self::WithFields(_) => {
+                // It's not possible to implement this without changint the return type, and some
+                // performance-critical VM serialization code uses the Runtime case of this.
+                // panicking is the best move
+                unimplemented!("Getting fields for decorated representation")
```
Comment (vgao1996) [inline language/move-core/types/src/value.rs:135]: `panic!` instead of `unimplemented!`?

File: language/move-core/types/src/value.rs [hunk 4]
```diff
@@ -72,28 +86,62 @@ where
...
     }
 
     pub fn fields(&self) -> &[MoveValue] {
-        &self.0
+        match self {
+            Self::Runtime(vals) => vals,
+            Self::WithFields(_) => {
+                // It's not possible to implement this without changing the return type, and some
+                // panicking is the best move
+                unimplemented!("Getting fields for decorated representation")
```
Comment (vgao1996) [inline language/move-core/types/src/value.rs:106]: `panic!` instead of `unimplemented!`?

File: language/move-core/types/src/value.rs (hunk 2 above)
Comment (sblackshear) [reply language/move-core/types/src/value.rs:223]: Good point. I'll go with the vec here.

**Review by vgao1996 — DISMISSED** [review]

**Review by vgao1996 — DISMISSED** [review]

**Pull request closed without merging**